subprojects/core/src/main/java/org/gradle/execution/plan/ExecutionNodeAccessHierarchy.java

```diff
@@ -0,0 +1,235 @@
+/*
+ * Copyright 2020 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.execution.plan;
+
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableSet;
+import org.gradle.api.file.FileTreeElement;
+import org.gradle.api.file.RelativePath;
+import org.gradle.api.specs.Spec;
+import org.gradle.execution.plan.ValuedPathHierarchy.ValueVisitor;
+import org.gradle.internal.UncheckedException;
+import org.gradle.internal.file.Stat;
+import org.gradle.internal.snapshot.CaseSensitivity;
+import org.gradle.internal.snapshot.EmptyChildMap;
+import org.gradle.internal.snapshot.VfsRelativePath;
+
+import java.io.File;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.nio.file.Files;
+import java.util.function.Supplier;
+
+public class ExecutionNodeAccessHierarchy {
+    private volatile ValuedPathHierarchy<NodeAccess> root;
+    private final Stat stat;
+
+    public ExecutionNodeAccessHierarchy(CaseSensitivity caseSensitivity, Stat stat) {
+        this.root = new ValuedPathHierarchy<>(ImmutableList.of(), EmptyChildMap.getInstance(), caseSensitivity);
+        this.stat = stat;
+    }
+
+    public ImmutableSet<Node> getNodesAccessing(String location) {
+        return visitValues(location, new AbstractNodeAccessVisitor() {
+            @Override
+            public void visitChildren(Iterable<NodeAccess> values, Supplier<String> relativePathSupplier) {
+                values.forEach(this::addNode);
+            }
+        });
+    }
+
+    public ImmutableSet<Node> getNodesAccessing(String location, Spec<FileTreeElement> filter) {
+        return visitValues(location, new AbstractNodeAccessVisitor() {
+            @Override
+            public void visitChildren(Iterable<NodeAccess> values, Supplier<String> relativePathSupplier) {
+                String relativePathFromLocation = relativePathSupplier.get();
+                if (filter.isSatisfiedBy(new ReadOnlyFileTreeElement(new File(location + "/" + relativePathFromLocation), relativePathFromLocation, stat))) {
+                    values.forEach(this::addNode);
+                }
+            }
+        });
+    }
```

Comment:
I think a bit of Javadoc on the public methods would be really useful.

Reply:
I added some Javadoc.